### ISSUE TITLE
use_entities is broken when domain is split

### ISSUE THREAD

**Issue by hsm207** [RasaHQ/rasa#10137]
### Rasa Open Source version

2.8.12

### Rasa SDK version

2.8.2

### Python version

3.8

### What operating system are you using?

Linux

### What happened?

Create a domain.yml with the following content:

```yaml
intents:
  - greet:
      use_entities: [country]
  - goodbye:
      use_entities: []
  - affirm
  - deny
  - mood_great
  - mood_unhappy
  - bot_challenge
 
entities:
  - city
  - country
```

Create another domain yaml with the following content:

```yaml
intents:
  - inform:
      use_entities: [city]
```

Reading the merged domain files using `Domain.load` shows  that the intent `inform` is still using the entities `city` and `country`.

The workaround is to repeat the list of entities in the first domain file in the second domain file i.e.:

```
intents:
  - inform:
      use_entities: [city]

entities:
  - city
  - country
```

Here is the [code](https://github.com/hsm207/rasa_moodbot/tree/bug_split_domain) to reproduce this behavior.

**Definition of done**
- [ ] Fix bug in 2.8 and 3.0
- [ ] Create integration test to ensure this works as expected
- [ ] Fixed merged into 2.8 and 3.0

**Comment by TyDunn** [comment]
The fix for this issue will need to be backported to the last minor of the previous major (i.e. Rasa Open Source 2.8)

**Comment by raoulvm** [comment]
I made a showcase on a possible solution here: https://github.com/raoulvm/rasa/commit/17b4d811bcfbc0d34a38f41b5d3d1419ef3ff632 in branch https://github.com/raoulvm/rasa/tree/2.8.14-fix-domain_from_directory
based on Rasa 2.8.12. It is **no complete PR** but a showcase which works with our (Deutsche Telekom) model.

**Comment by TyDunn** [comment]
@raoulvm Awesome. Thank you so much! This will help us make progress quickly on this issue

**Comment by carlad** [comment]
@raoulvm am working on this right now. I've identified the root cause but need to ensure my fix doesn't break other things ;) I'll message you if I need some help.

**Comment by carlad** [comment]
@TyDunn I've raised a [PR](https://github.com/RasaHQ/rasa/pull/10424) for `2.8.x` and added a unit test. I don't think we need an integration test as the bug was isolated to a quirk in one of the `shared.core.domain` class methods only...

**Comment by raoulvm** [comment]
I have a bad feeling.  
@carlad could you please test with the following two configs:

```YAML
intents:
  - play:
      use_entities:
        - ball
        - chess
 - stow_away:
     use_entities: true
entities:
  - ball
  - chess
 ```

```YAML
intents:
  - support_banning:
      use_entities:
         - automatic_rifles
         - anti_person_land_mines
  - certify:
      use_entities: true
entities:
  - automatic_rifles
  - anti_person_land_mines
  - tanks
```
and post the results here?  
Thanks

**Comment by raoulvm** [comment]
Spoiler, that is expected, when using `domain.as_yaml(clean_before_dump=True)`:
```
version: '2.0'
session_config:
  session_expiration_time: 60
  carry_over_slots_to_new_session: true
intents:
- support_banning:
    use_entities:
    - automatic_rifles
    - anti_person_land_mines
- certify
- play:
    use_entities:
    - ball
    - royal_chess
- stow_away
entities:
- anti_person_land_mines
- automatic_rifles
- ball
- royal_chess
- tanks
```

**Comment by carlad** [comment]
This is what I get when I use your `rasa_moodbot` repo with this PR's fix:

```
❯ make check-bug
python main.py --domain_folder ./domain/bug
intent: certify
entities used:
  anti_person_land_mines
  automatic_rifles
  ball
  chess
  tanks


intent: play
entities used:
  anti_person_land_mines
  automatic_rifles
  ball
  chess
  tanks


intent: stow_away
entities used:
  anti_person_land_mines
  automatic_rifles
  ball
  chess
  tanks


intent: support_banning
entities used:
  anti_person_land_mines
  automatic_rifles
  ball
  chess
  ```

**Comment by carlad** [comment]
Are you able to raise a PR from your branch against `2.8.x`

**Comment by raoulvm** [comment]
You could just copy the code changes. I provided that as an example.

**Comment by raoulvm** [comment]
There are open topics though: If the domains are _not_ loaded from a directory, but individually listed in CLI, the fix does not apply. The change probably would be similar, but I neither had time nor interest to tackle that.

**Comment by carlad** [comment]
You linked to the entire branch. Which files did you change?

**Comment by raoulvm** [comment]
Let me have quick look at the commits.

**Comment by raoulvm** [comment]
https://github.com/raoulvm/rasa/commit/17b4d811bcfbc0d34a38f41b5d3d1419ef3ff632
That commit contains the changes.

**Comment by carlad** [comment]
@raoulvm thanks! Let me add your changes to this PR...

**Comment by carlad** [comment]
@raoulvm this fix has now been released in `2.8.19`: https://pypi.org/project/rasa/2.8.19/
